gateleen-routing/src/test/java/org/swisspush/gateleen/routing/ForwarderTest.java

```diff
@@ -0,0 +1,132 @@
+      rule.setHeaderFunction(headerFunction);
+    }
+    rule.setUrlPattern(extractedRule.getString("url"));
+    rule.setHost(DEFAULTHOST);
+    rule.setPort(DEFAULTPORT);
+    return rule;
+  }
+
+  @Before
+  public void setUp() {
+    vertx = Mockito.mock(Vertx.class);
+    loggingResourceManager = Mockito.mock(LoggingResourceManager.class);
+    monitoringHandler = Mockito.mock(MonitoringHandler.class);
+    httpClient = Mockito.mock(HttpClient.class);
+    storage = new MockResourceStorage(ImmutableMap.of(RULES_PATH, RULES));
+    logger = LoggerFactory.getLogger(ForwarderTest.class.getName());
+  }
+
+  @Test
+  public void testHeaderFunctionsWithHostHeader() {
+    Rule rule = extractRule("/ruleWithHostHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, "oldHost1234");
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),"newHost");
+  }
+
+  @Test
+  public void testHeaderFunctionsDefaultWithHostHeader() {
+    Rule rule = extractRule("/ruleWithHostHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, DEFAULTHOST + ":" + DEFAULTPORT);
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),"newHost");
+  }
+
+  @Test
+  public void testHeaderFunctionsWithoutHostHeader() {
+    Rule rule = extractRule("/ruleWithoutHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, "oldHost:1234");
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),DEFAULTHOST + ":" + DEFAULTPORT);
+  }
+
+  @Test
+  public void testHeaderFunctionsDefaultWithoutHostHeader() {
+    Rule rule = extractRule("/ruleWithoutHeader");
+    Forwarder forwarder = new Forwarder(vertx, httpClient, rule, storage, loggingResourceManager,
+        monitoringHandler, USER_PROFILE_PATH);
+    MultiMap reqHeaders = new VertxHttpHeaders();
+    reqHeaders.add(HOST_HEADER, DEFAULTHOST + ":" + DEFAULTPORT);
+    String errorMessage = forwarder.applyHeaderFunctions(logger, reqHeaders);
+    Assert.assertNull(errorMessage);
+    Assert.assertEquals(reqHeaders.get(HOST_HEADER),DEFAULTHOST + ":" + DEFAULTPORT);
+  }
+
+
+}
```

Comment:
*trivial*: I think would be nice to have a test which would test the "*return-an-error-string*" case. But I had no simple idea how to "*just-write*" such a test.

Reply:
Did not find a way to test this properly and to force failing the evaluation chain.